### ISSUE TITLE
runtime: fatal error: string concatenation too long

### ISSUE THREAD

**Issue by mikioh** [golang/go#10890]
On linux/amd64 with 0b9866fd561d5ecebee14d73c0a4938dec5abe7d. I saw this only once during a test like the following:
```
#!/usr/bin/env bash

# go test -c

set -xe
while true; do
        GOMAXPROCS=$[ 1 + $[ RANDOM % 128 ]] ./net.test -test.v=true -test.short=false -test.run=TestVariousDeadlines || exit 255
done
```

**Comment by minux** [comment]
stack trace?

**Comment by mikioh** [comment]
The dying message:
```
timeout_test.go:811: 1s run 3/3
        timeout_test.go:836: for 1s run 3/3, good client timeout after 1.02473192s, reading 34086912 bytes
        timeout_test.go:846: for 1s run 3/3, server in 1.024844774s wrote 35192832: read tcp4 127.0.0.1:44148->127.0.0.1:50913: write tcp4 127.0.0.1:44148->127.0.0.1:50913: write: connection reset by peer
=== RUN   TestVariousDeadlines4Proc-96
fatal error: string concatenation too long

goroutine 1095 [running]:
runtime.throw(0x8119b0, 0x1d)
        /go/src/runtime/panic.go:527 +0x96 fp=0xc208467040 sp=0xc208467028
runtime.concatstrings(0x0, 0xc2084671a0, 0x3, 0x3, 0x0, 0x0)
        /go/src/runtime/string.go:32 +0x38d fp=0xc208467148 sp=0xc208467040
runtime.concatstring3(0x0, 0xc208119220, 0xdeaddeaddeaddead, 0x7c2690, 0x2, 0x7dbc70, 0xb, 0x0, 0x0)
        /go/src/runtime/string.go:62 +0x71 fp=0xc208467198 sp=0xc208467148
os.(*SyscallError).Error(0xc208119200, 0x0, 0x0)
        /go/src/os/error.go:34 +0xe5 fp=0xc208467208 sp=0xc208467198
net.(*OpError).Error(0xc2084d2af0, 0x0, 0x0)
        /go/src/net/net.go:376 +0x418 fp=0xc2084672a8 sp=0xc208467208
net.(*OpError).Error(0xc2084d2b40, 0x0, 0x0)
        /go/src/net/net.go:376 +0x418 fp=0xc208467348 sp=0xc2084672a8
fmt.(*pp).handleMethods(0xc20824a000, 0xc200000076, 0x0, 0x1)
        /go/src/fmt/print.go:719 +0x5ca fp=0xc208467428 sp=0xc208467348
fmt.(*pp).printArg(0xc20824a000, 0x792da0, 0xc2084d2b40, 0x76, 0x0, 0x0)
        /go/src/fmt/print.go:801 +0x5d3 fp=0xc2084675b0 sp=0xc208467428
fmt.(*pp).doPrintf(0xc20824a000, 0x814e70, 0x21, 0xc208467dd0, 0x4, 0x4)
        /go/src/fmt/print.go:1197 +0x2c4a fp=0xc2084679a8 sp=0xc2084675b0
fmt.Sprintf(0x814e70, 0x21, 0xc208467dd0, 0x4, 0x4, 0x0, 0x0)
        /go/src/fmt/print.go:203 +0x80 fp=0xc2084679f8 sp=0xc2084679a8
testing.(*common).Logf(0xc20801c090, 0x814e70, 0x21, 0xc208467dd0, 0x4, 0x4)
        /go/src/testing/testing.go:354 +0x5e fp=0xc208467a48 sp=0xc2084679f8
net.testVariousDeadlines(0xc20801c090, 0x4)
        /go/src/net/timeout_test.go:846 +0x147b fp=0xc208467f70 sp=0xc208467a48
net.TestVariousDeadlines4Proc(0xc20801c090)
        /go/src/net/timeout_test.go:726 +0x3f fp=0xc208467f88 sp=0xc208467f70
testing.tRunner(0xc20801c090, 0x967348)
        /go/src/testing/testing.go:454 +0xfc fp=0xc208467fd0 sp=0xc208467f88
runtime.goexit()
        /go/src/runtime/asm_amd64.s:1670 +0x1 fp=0xc208467fd8 sp=0xc208467fd0
created by testing.RunTests
        /go/src/testing/testing.go:562 +0xc96

goroutine 1 [chan receive]:
testing.RunTests(0x85dff8, 0x966700, 0x95, 0x95, 0x956801)
        /go/src/testing/testing.go:563 +0xcec
testing.(*M).Run(0xc2080b9f40, 0x0)
        /go/src/testing/testing.go:492 +0xe8
net.TestMain(0xc2080b9f40)
        /go/src/net/main_test.go:44 +0x40
main.main()
        net/_test/_testmain.go:396 +0x20d

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
        /go/src/runtime/asm_amd64.s:1670 +0x1

goroutine 1096 [IO wait]:
net.runtime_pollWait(0x7fd6421ee5e0, 0x72, 0x7fd6421e9028)
        /go/src/runtime/netpoll.go:157 +0x66
net.(*pollDesc).Wait(0xc208524a00, 0x72, 0x0, 0x0)
        /go/src/net/fd_poll_runtime.go:73 +0x5d
net.(*pollDesc).WaitRead(0xc208524a00, 0x0, 0x0)
        /go/src/net/fd_poll_runtime.go:78 +0x4b
net.(*netFD).accept(0xc2085249a0, 0x0, 0x7fd6421ee860, 0xc2081191e0)
        /go/src/net/fd_unix.go:459 +0x2f9
net.(*TCPListener).AcceptTCP(0xc2084702b0, 0xc2080bbf50, 0x0, 0x0)
        /go/src/net/tcpsock_posix.go:254 +0x7a
net.(*TCPListener).Accept(0xc2084702b0, 0x0, 0x0, 0x0, 0x0)
        /go/src/net/tcpsock_posix.go:264 +0x4e
net.testVariousDeadlines.func1(0xc208474030, 0x7fd6421ed638, 0xc2084702b0)
        /go/src/net/timeout_test.go:756 +0x51
net.(*localServer).buildup.func1(0xc2085c5120, 0xc208474030)
        /go/src/net/mockserver_test.go:85 +0x6b
created by net.(*localServer).buildup
        /go/src/net/mockserver_test.go:87 +0x60
+ exit 255
```
Looks like the os.SyscallError is corrupted.

**Comment by mikioh** [comment]
```
runtime.concatstring3(0x0, 0xc208119220, 0xdeaddeaddeaddead, 0x7c2690, 0x2, 0x7dbc70, 0xb, 0x0, 0x0)

type SyscallError struct {
        Syscall string
        Err     error
}

func (e *SyscallError) Error() string { return e.Syscall + ": " + e.Err.Error() }
```
SyscallError.Syscall = (0xc208119220, 0xdeaddeaddeaddead)
": " = (0x7c2690, 0x2)
SyscallError.Err = "broken pipe" = (0x7dbc70, 0xb)

**Comment by minux** [comment]
Indeed, apparently a memory corruption.

Could you try bisect the commit that causes this?

**Comment by mikioh** [comment]
I just tried to reproduce this issue on the same instance, run the script 10mins, but didn't get any crashes. Perhaps it might require a bit long-run to repro.

**Comment by rsc** [comment]
Duplicate of #11116.
